### ISSUE TITLE
-short-paths incorrectly picks arbitrary type alias

### ISSUE THREAD

**Issue by mjambon** [ocaml/ocaml#10432]
Quick reproduction with ocaml 4.10.0 or 4.12.0: visit and clone https://github.com/mjambon/short-paths-alias-bug then run `make`.

Description
--

When an ml file contains a type definition and one or more aliases, errors involving those types are reported using one of the aliases instead of the name bound in the original type definition. This results in confusing error messages as well as confusing inferred types showed by merlin.

It occurs as a combination of the following:
* when referencing a type defined in another module (or compilation unit)
* when using `-short-paths` (the default with dune)

It appears the alias chosen for the error message is the last one by alphabetic order. In the example below, it is `t_ee` but we want `t_c`, or perhaps whichever alias is used as type annotation/constraint. Since we're not indicating a preference for type `t_ee` anywhere, it should not occur in the error messages.

File `Debug1.ml`:
```ocaml
(* original type definition *)
type t_c = A

(* aliases *)
type t_b = t_c
type t_d = t_c
type t_ee = t_c
type t_a = t_c

let x = A
```

File `Debug2.ml`:
```ocaml
open Debug1

(* error message reports type 't_ee' instead of 't_c' *)
let y = x + 1
```

File `Debug3.ml`:
```ocaml
(* error message reports type 'Debug1.t_ee' instead of 'Debug1.t_c' *)
let y = Debug1.x + 1
```

Output:
```
$ make
ocamlc -c Debug1.ml
=== shows the correct type alias ===
ocamlc -c Debug2.ml
File "Debug2.ml", line 4, characters 8-9:
4 | let y = x + 1
            ^
Error: This expression has type Debug1.t_c
       but an expression was expected of type int
make: [Makefile:11: build] Error 2 (ignored)
echo

=== shows the wrong type alias ===
ocamlc -c -short-paths Debug2.ml
File "Debug2.ml", line 4, characters 8-9:
4 | let y = x + 1
            ^
Error: This expression has type t_ee but an expression was expected of type
         int
make: [Makefile:14: build] Error 2 (ignored)
echo

=== shows the wrong type alias ===
ocamlc -c -short-paths Debug3.ml
File "Debug3.ml", line 2, characters 8-16:
2 | let y = Debug1.x + 1
            ^^^^^^^^
Error: This expression has type Debug1.t_ee
       but an expression was expected of type int
make: [Makefile:17: build] Error 2 (ignored)
```

**Comment by Octachron** [comment]
This is the intended effect of `-short-paths`: it picks a common "best" paths amongst all aliases in scope.

A illustration of the intended behavior would be:
```ocaml
module A =  struct type t_c = A
  (* aliases *)
  type t_b = t_c
  type t_d = t_c
  type t_a = t_c
  type t_ee = t_c
end
open A
module M: sig
  val f: t_c -> t_b -> t_a -> t_ee -> int
end = struct
  let f: t_b -> t_a -> t_ee -> t_c -> float = fun _ _ _ _ -> 0.
end
```
The error message with `-short-paths` is
>```Error: Signature mismatch:
>       Modules do not match:
>         sig val f : t_ee -> t_ee -> t_ee -> t_ee -> float end
>       is not included in
>         sig val f : t_ee -> t_ee -> t_ee -> t_ee -> int end
>       Values do not match:
>         val f : t_ee -> t_ee -> t_ee -> t_ee -> float
>       is not included in
>         val f : t_ee -> t_ee -> t_ee -> t_ee -> int
>```
compared to

>```       Modules do not match:
>        sig val f : A.t_b -> A.t_a -> A.t_ee -> A.t_c -> float end
>      is not included in
>        sig val f : A.t_c -> A.t_b -> A.t_a -> A.t_ee -> int end
>      Values do not match:
>        val f : A.t_b -> A.t_a -> A.t_ee -> A.t_c -> float
>      is not included in
>         val f : A.t_c -> A.t_b -> A.t_a -> A.t_ee -> int
>```
without.

Nevertheless, we could tweak the heuristic for the choice of this "best" name to pick the first name in alphabetical order.

**Comment by mjambon** [comment]
> This is the intended effect of -short-paths: it picks a common "best" paths amongst all aliases in scope.

I don't think there's an issue of path here. We would like a preference for the original name over any of its aliases **defined in the same module**; not over any aliases in general. This seems compatible with the behavior of `-short-paths`.

I'm not sure how this is implemented and it might be trickier than I imagine... but I picture two paths that differ only by their last component e.g. `A.B.original` and `A.B.alias`. The first one's priority would be (3, 0) and the second one's priority would be (3, 1), where the first number is the path length the second number is a tie-breaker that indicates whether the name is an alias.

> Nevertheless, we could tweak the heuristic for the choice of this "best" name to pick the first name in alphabetical order.

This would be better than nothing I guess, but it would require naming our aliases with an `zz_` prefix. Currently, the workaround is to use `aa_` as a prefix.

We cannot place the aliases into a submodule (so as to create a longer path) because the type definitions and alias definitions are mutually recursive. For illustration, we're dealing with [this](https://github.com/returntocorp/pfff/blob/2f3c1ce703754c01e0ec7a54f21e2e4c95be109a/lang_js/parsing/ast_js.ml) and other ASTs, in which we prefer `expr` over any of its aliases.

**Comment by Gbury** [comment]
> I don't think there's an issue of path here. We would like a preference for the original name over any of its aliases **defined in the same module**; not over any aliases in general. This seems compatible with the behavior of `-short-paths`.

That might not always be the case: for instance, when dealing with format types (cf https://ocaml.org/api/Stdlib.html#1_Operationsonformatstrings ), it might be nicer to keep aliases such as `format4` and `format`, rather than always expanding to `format6`.

**Comment by mjambon** [comment]
@Gbury good point.

`format4` is a specialized version of `format6`:
```ocaml
type ('a, 'b, 'c, 'd) format4 = ('a, 'b, 'c, 'c, 'c, 'd) format6
```
It makes sense to prefer the specialized version, `format4`, which is kind of an "alias" for the original type `format6` but not really. I'm not sure about the terminology but `format4` is not equivalent to `format6` in the sense of representing the same set of values.

In the case of a "true alias" representing the same set:
```ocaml
type 'a long_name = A of 'a
type 'a t = 'a long_name
```
we would have a tie, and preference should be given to the original name `long_name` rather than the alias. If the author thinks `t` is the canonical name, they have control over this. They would write their code as:
```ocaml
type 'a t = A of 'a
type 'a long_name = 'a t
```
And this is predictable for both the author of the module and for the user.

There are interesting cases like these:
```ocaml
type ('a, 'b) ab = X of 'a * 'b
type ('a, 'b) ba = ('b, 'a) ab
```
^ Ideally, those should be treated as aliases, under the rule that they represent the same set of values. Maybe determining this equivalence is complicated. Since this is not a common situation at all, it doesn't matter much how we treat those.

In summary, it looks like we could stick the "is an alias" property to a type declaration/definition iff the right-hand side has the same structure as the left-hand side, including identical type parameters.

Examples of aliases:
```ocaml
type a = b
type ('a, 'b) long = ('a, 'b) t
```
Examples of non-aliases for our purpose:
```ocaml
type t = int orig
type 'a constrained = ('a, 'a) orig
type ('a, 'b) riap = ('b, 'a) pair
type 'a t = orig
type 'a t = 'a orig constraint 'a = int
```

I assume this is all for the purpose of showing better type names when there's a choice. The notion of alias used for this purpose is part of the heuristic and would remain subject to change.

Also, I can't tell if this should be part of `-short-paths` or if it should be an independent option.

**Comment by Octachron** [comment]
The distinction between transparent and not transparent aliases is already implemented.

Currently, `-short-paths` merely does not discriminate between transparent aliases and the "original" definition.

It is not that clear if the original name is always the best choice. For instance, a type might need to be first defined in a mutually recursive group in a `Type_definitions` module, whereas its canonical representation lives in the module `M`. If you want to have `hidden` aliases from the point of view of `-short-paths`, you can insert `__` in their name.

An important point to note is that we are currently  planning to switch to Merlin's version of `-short-paths`.

**Comment by mjambon** [comment]
> It is not that clear if the original name is always the best choice. For instance, a type might need to be first defined in a mutually recursive group in a `Type_definitions` module, whereas its canonical representation lives in the module `M`.

This makes sense. The original problem I was reporting occurs only when two or more aliases are defined in the same module. Here's a tentative selection rule:
1. Consider the tree of aliases.
2. Start from the subset _S_ of aliases that are tied after eliminating long paths like `Long.Long.Path.t`.
3. Filter _S_ to keep only the aliases with no descendants in that set, giving _S'_. For example, if there's `type b = a` and both `a` and `b` are in _S_, prefer `b`.
4. Find the most recent common ancestor of _S'_ and use that.

The goal of (2) is to focus on only short paths, eliminating longer aliases that might be defined in submodules. Step (4) guarantees a unique solution. This solution may be outside of _S_, i.e. it may have a longer path than the original candidates (discussed later). Step (3) is a preliminary for step (4) and sets a preference for the latest alias in a lineage. The goal of step (4) is then to return the most recent non-confusing name.

Here's an example with sibling/cousin aliases, similar to the original problem I posted:

Module `A.B` (long path, not desired):
```ocaml
type t
```

Current module:
```ocaml
type t = A.B.t
type expr = t
type pattern = t
```

The tree of type aliases is:
```
A.B.t
└── t
    ├── expr
    └── pattern
```
Say the short-paths heuristic discards `A.B.t` because it's too long. We're left with a tie between `t`, `expr`, and `pattern` (set _S_). Step (3) discards `t`, leaving _S'_ = {`expr`, `pattern`}. Their most recent ancestor is `t`, which is selected as the preferred name.

Here's an example with a single alias:
```
A.B.t
└── expr
    └── t
```
In that case, the candidates are `expr` and `t`. `expr` is eliminated because `t` is a more recent candidate in the same lineage (step 3). The solution is `t`.

Here's a case where the solution has a longer path than the original candidates:
```
type cells = Tbl.t
type objects = Tbl.t
```
The tree is:
```
Tbl.t
├── cells
└── objects
```
In that case, the selected name is `Tbl.t` rather than `cells` or `objects`. I think this is more desirable than arbitrarily picking `cells` or `objects`. This situation can be easily avoided by the programmer by creating a `container` alias and write `type cells = container`:
```
Tbl.t
└── container
    ├── cells
    └── objects
```
Now, the preferred name is `container`.

I think the most important aspect here is to give control to the programmer over what name will be selected. Having to write extra code such as adding new names or using special annotations is totally fine. Having to change the names we care about is not great, though.

**Comment by mjambon** [comment]
> An important point to note is that we are currently planning to switch to Merlin's version of `-short-paths`.

Should this issue be forwarded to the merlin project?

**Comment by Octachron** [comment]
I forgot to mention my main point: I am not sure if anyone will have the time for tuning the `-short-paths` implementation of either Merlin or the compiler before those two variants are synchronized.

**Comment by mjambon** [comment]
No worries, we're not blocked. We use the [`a_` prefix workaround](https://github.com/returntocorp/pfff/pull/433/files).

**Comment by gasche** [comment]
Is fine-tuning the heuristic really the correct way to go here? @mjambon it sounds like you use type aliases to convey important distinctions, not just shorten definitions. This is fragile, and will always be, because printing aliases will always contain some arbitrariness. What about not using aliases there? Would it be unworkable in practice to use, say, `type bar = Foo of foo` instead? (For example, `type pat = Pat of expr` instead of `type pat = expr`.)

**Comment by mjambon** [comment]
@gasche as you suggest, `type pat = Pat of expr` would work but our application is performance-sensitive and this solution worries me.

Maybe we could use phantom types. We usually try to keep things simple, but this may be worth it:
```ocaml
type 'a expr_like = A

(* abstract types used only as type parameters *)
type expr
type pat
type xml

(* aliases used for writing type annotations; it doesn't matter if the type checker doesn't use them *)
type expression = expr expr_like
type pattern = pat expr_like
type xml_attr_value = xml expr_like
```

The following gives a type error, which is good:
```ocaml
let f (p : pattern) (e : expression) =
  p = e
```
```
Error: This expression has type expression = expr expr_like
       but an expression was expected of type pattern = pat expr_like
       Type expr is not compatible with type pat 
```

**Comment by Octachron** [comment]
If the problem is just performance, with an unboxed annotation
```ocaml
type pat = Pat of expr [@@unboxed]
```
the constructor will not exist after type checking.

**Comment by mjambon** [comment]
@Octachron I didn't know about this feature. Very nice!
